### PR TITLE
Support other flavour of CRI

### PR DESCRIPTION
**Pull request by CiraciNicolo**
## Change description

This PR implements support for different CRI implementations, instead of the default containerd. At this moment the only other CRI implementation supported is CRI-O. By default containerd is still the default CRI, to start using CRI-O references the packer configuration called "qemu-ubuntu-2204-crio.json".

Currently the only operating systems supported are debian based and centos based. It should also work on rhel but not tested.

The implementation can be done following this steps:
- create an ansible role for your CRI flavour
- implement default packer variables
- create builder target and makefile target

Both ansible role and default values should be called as your flavour, in this way you can set the ```cri_flavour``` build variable and all tasks/var should be included.

Be also careful to correctly default the variables in ```images/capi/packer/config/ansible-args.json``` and implement the goss specs.

## Related issues
No issues are addressed by this PR

This PR also includes the following fixes:
- Fix incorrect boolean evalutation for GVisor installation if enabled. Be careful that this fixes could change the expected behaviour if you are leveraging Gvisor

### CODE REVIEW

File: images/capi/packer/qemu/qemu-ubuntu-2004.json [hunk 1]
```diff
@@ -10,4 +10,4 @@
   "iso_url": "http://cdimage.ubuntu.com/ubuntu-legacy-server/releases/20.04/release/ubuntu-20.04.1-legacy-server-amd64.iso",
   "os_display_name": "Ubuntu 20.04",
   "shutdown_command": "shutdown -P now"
-}
+}
```
Comment (mboersma) [inline images/capi/packer/qemu/qemu-ubuntu-2004.json:13]: Could you remove this diff? (That is, keep the final newline.)

File: images/capi/packer/qemu/qemu-ubuntu-2204-crio.json [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+  "distribution_version": "2204",
+  "distro_name": "ubuntu",
+  "guest_os_type": "ubuntu-64",
+  "iso_checksum": "45f873de9f8cb637345d6e66a583762730bbea30277ef7b32c9c3bd6700a32b2",
+  "iso_checksum_type": "sha256",
+  "iso_url": "https://old-releases.ubuntu.com/releases/jammy/ubuntu-22.04.4-live-server-amd64.iso",
+  "os_display_name": "Ubuntu 22.04",
+  "shutdown_command": "shutdown -P now",
+  "unmount_iso": "true"
+}
```
Comment (mboersma) [inline images/capi/packer/qemu/qemu-ubuntu-2204-crio.json:15]: Could you add a final newline here?

File: images/capi/packer/qemu/qemu-ubuntu-2204-efi.json [hunk 3]
```diff
@@ -11,4 +11,4 @@
   "os_display_name": "Ubuntu 22.04",
   "shutdown_command": "shutdown -P now",
   "unmount_iso": "true"
-}
+}
```
Comment (mboersma) [inline images/capi/packer/qemu/qemu-ubuntu-2204-efi.json:14]: Remove this diff please / keep final newline.

File: images/capi/packer/qemu/qemu-ubuntu-2204.json [hunk 4]
```diff
@@ -10,4 +10,4 @@
   "os_display_name": "Ubuntu 22.04",
   "shutdown_command": "shutdown -P now",
   "unmount_iso": "true"
-}
+}
```
Comment (mboersma) [inline images/capi/packer/qemu/qemu-ubuntu-2204.json:13]: Remove this diff please / keep final newline.

File: images/capi/ansible/roles/crio/templates/etc/crio/crio.conf [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+conmon_cgroup = "pod"
+cgroup_manager = "cgroupfs"
+selinux = false
+
+[crio.runtime.runtimes.gvisor]
+runtime_path = "/usr/local/bin/runsc"
+runtime_type = "oci"
+{% endif %}
+
+{{crio_additional_settings | b64decode}}
```
Comment (mboersma) [inline images/capi/ansible/roles/crio/templates/etc/crio/crio.conf:17]: Please add a final newline.

File: images/capi/ansible/roles/crio/tasks/redhat.yml [hunk 6]
```diff
@@ -0,0 +1,46 @@
+# Copyright 2018 The Kubernetes Authors.
```
Comment (mboersma) [inline images/capi/ansible/roles/crio/tasks/redhat.yml:1]:
```suggestion
# Copyright 2024 The Kubernetes Authors.
```

My understanding is that new files (even if they are adapted from elsewhere in the project) assert the current copyright year, so all new files here should have "2024".

File: docs/book/src/capi/cri/cri.md [hunk 7]
```diff
@@ -0,0 +1,36 @@
+# Container Runtime
+
+The image-builder project support different implementation, referred as flavour in this book, as CRI. The preferred one is containerd but cri-o is supported for some kind of platforms - depending on cri-o supported operating systems. By default the built-in CNI provided by CRI-O is disabled, to enabled use the following variable:
```
Comment (mboersma) [inline docs/book/src/capi/cri/cri.md:3]:
```suggestion
The image-builder project supports different implementations of CRI, referred to as flavours in this book. The preferred option is containerd, but CRI-O is supported for some platforms. By default, the built-in CNI provided by CRI-O is disabled. To enable it use the following variable:
```

File: docs/book/src/capi/cri/cri.md [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+```json
+{
+  "crio_disable_default_cni": "false",
+}
+```
+
+## crictl
+
+When cri-o is preferred, crictl is not provided with the package. This means you need to install it using the http source type:
```
Comment (mboersma) [inline docs/book/src/capi/cri/cri.md:13]:
```suggestion
When CRI-O is specified, `crictl` is not provided with the installation package. You need to install it using the `http` source type:
```

File: docs/book/src/capi/cri/cri.md [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+```json
+{
+  "crictl_source_type": "http",
+}
+```
+
+## Running sandboxed containers using gVisor
+
+As of now gVisor support is implemented but broken - [gvisor/issue/3283](https://github.com/google/gvisor/issues/3283). Refer to the [relative section](./customizing-containerd.md) keeping in mind that the variables are similar but different:
```
Comment (mboersma) [inline docs/book/src/capi/cri/cri.md:23]:
```suggestion
As of now, gVisor support is implemented but broken. See [gvisor/issue/3283](https://github.com/google/gvisor/issues/3283). Refer to the [relative section](./customizing-containerd.md), keeping in mind that the variables are similar but different:
```

File: docs/book/src/capi/cri/cri.md [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+
+```json
+{
+    "crio_gvisor_runtime": true,
+    "crio_gvisor_version": "yyyymmdd", // or latest
```
Comment (mboersma) [inline docs/book/src/capi/cri/cri.md:28]:
```suggestion
    "crio_gvisor_version": "yyyymmdd", // or "latest"
```

File: docs/book/src/capi/cri/cri.md [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+}
+```
+
+For example you can build the qcow2 image with gvisor enabled with this snippet:
```
Comment (mboersma) [inline docs/book/src/capi/cri/cri.md:32]:
```suggestion
For example, you can build the qcow2 image with gVisor enabled with this command:
```

**Review by mboersma — COMMENTED** [review]
Great stuff--this is a lot of work!

I did a quick review and just had some documentation suggestions. Also two general comments:
- please update the copyright year to `2024` in new files
- please add a final newline to all files

File: images/capi/packer/goss/goss-vars.yaml [hunk 12]
```diff
@@ -97,6 +97,8 @@ photon_5_rpms: &photon_5_rpms
   jq:
 
 arch: "amd64"
+cri_flavour: "containerd"
+crio_version: ""
 containerd_version: ""
```
Comment (jsturtevant) [inline images/capi/packer/goss/goss-vars.yaml:102]: thought: would be it be easier if this was just `cri_version`? We would need to keep `containerd_version` around for a release but then could remove it in a future release?

File: images/capi/ansible/roles/crio/tasks/main.yml [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+---
+- ansible.builtin.import_tasks: debian.yml
+  when: ansible_os_family == "Debian"
+
+- ansible.builtin.import_tasks: redhat.yml
+  when: ansible_os_family == "RedHat"
+
+- name: Export crisocket
+  set_fact:
+    cri_socket: "{{ crio_cri_socket }}"
```
Comment (jsturtevant) [inline images/capi/ansible/roles/crio/tasks/main.yml:23]: Same with `crio_cri_socket`, should this be `cri_socket`?  And later we can remove the `containerd_cri_socket` from the public api?

**Review by jsturtevant — COMMENTED** [review]
Great work!  A few comments on the public API for cri specific fields

Do we have a way to test this in CI?

**Comment by CiraciNicolo** [comment]
Thanks! I will check all your reports tomorrow. Great idea about removing specific variables for containerd, I was afraid doing that would introduce some issues.

File: images/capi/packer/qemu/qemu-ubuntu-2004.json (hunk 1 above)
Comment (CiraciNicolo) [reply images/capi/packer/qemu/qemu-ubuntu-2004.json]: Should be OK

**Comment by jsturtevant** [comment]
> Great idea about removing specific variables for containerd, I was afraid doing that would introduce some issues.

To be clear we should keep the variables for a release or two and mark them depreciated, otherwise it would could some hickups for folks I think.

**Comment by CiraciNicolo** [comment]
Hi, since I messed up and worked on main instead of a branch I'm gonna close this PR and rework my commits into a dedicated branch. We can continue discuss there! Hope this is ok.

**Comment by CiraciNicolo** [comment]
Hi! I cherry picked the commits to another branch, this is more cleaner. Please, refer to #1442.

Sorry for the incorrect handling on the main branch.

**Pull request closed without merging**